reconcile/external_resources/manager.py

```diff
@@ -0,0 +1,330 @@
+                key=key,
+                resource_hash=state.resource_hash,
+                module_configuration=state.reconciliation.module_configuration,
+                input=state.reconciliation.input,
+                action=Action.DESTROY,
+            )
+            r.add(reconciliation)
+        return r
+
+    def _update_in_progress_state(
+        self, r: Reconciliation, state: ExternalResourceState
+    ) -> bool:
+        """Gets the resource reconciliation state from the Job and updates the
+        Resource state accordingly. It also returns if the target outputs secret needs
+        to be reconciled.
+
+        :param r: Reconciliation object
+        :param state: State object
+        :return: True/False if target Secret needs to be reconciled.
+        """
+        need_secret_sync = False
+
+        if state.resource_status not in set([
+            ResourceStatus.DELETE_IN_PROGRESS,
+            ResourceStatus.IN_PROGRESS,
+        ]):
+            return need_secret_sync
+
+        logging.info(
+            "Reconciliation In progress. Action: %s, Key:%s",
+            state.reconciliation.action,
+            state.reconciliation.key,
+        )
+
+        # Need to check the reconciliation set in the state, not the desired one
+        # as the reconciliation object might be from a previous desired state
+        error = False
+        match self.reconciler.get_resource_reconcile_status(state.reconciliation):
+            case ReconcileStatus.SUCCESS:
+                logging.info(
+                    "Reconciliation ended SUCCESSFULLY. Action: %s, key:%s",
+                    r.action.value,
+                    r.key,
+                )
+                if r.action == Action.APPLY:
+                    state.resource_status = ResourceStatus.CREATED
+                    state.reconciliation_errors = 0
+                    self.state_mgr.set_external_resource_state(state)
+                    need_secret_sync = True
+                elif r.action == Action.DESTROY:
+                    state.resource_status = ResourceStatus.DELETED
+                    self.state_mgr.del_external_resource_state(r.key)
+            case ReconcileStatus.ERROR:
+                logging.info(
+                    "Reconciliation ended with ERROR: Action:%s, Key:%s",
+                    r.action.value,
+                    r.key,
+                )
+                error = True
+            case ReconcileStatus.NOT_EXISTS:
+                logging.info(
+                    "Reconciliation should exist but it doesn't. Marking as ERROR to retrigger: Action:%s, Key:%s",
+                    r.action.value,
+                    r.key,
+                )
+                error = True
+        if error:
+            state.resource_status = ResourceStatus.ERROR
+            state.reconciliation_errors = state.reconciliation_errors + 1
+            self.state_mgr.set_external_resource_state(state)
+
+        return need_secret_sync
```

Comment:
`error` and `need_secret_sync` live too long in this method, make it harder to track changes, how about unify status handling and do early return?

```python
        match self.reconciler.get_resource_reconcile_status(state.reconciliation):
            case ReconcileStatus.SUCCESS:
                logging.info(
                    "Reconciliation ended SUCCESSFULLY. Action: %s, key:%s",
                    r.action.value,
                    r.key,
                )
                if r.action == Action.APPLY:
                    state.resource_status = ResourceStatus.CREATED
                    state.reconciliation_errors = 0
                    self.state_mgr.set_external_resource_state(state)
                    return True
                elif r.action == Action.DESTROY:
                    state.resource_status = ResourceStatus.DELETED
                    self.state_mgr.del_external_resource_state(r.key)
                    return False
            case (ReconcileStatus.ERROR | ReconcileStatus.NOT_EXISTS) as error_status:
                logging.info(
                    "Reconciliation ended with error status %s: Action:%s, Key:%s",
                    error_status,
                    r.action.value,
                    r.key,
                )
                state.resource_status = ResourceStatus.ERROR
                state.reconciliation_errors = state.reconciliation_errors + 1
                self.state_mgr.set_external_resource_state(state)
                return False
            case _:
                return False
```